internal/impl/git/input.go

```diff
@@ -0,0 +1,550 @@
+			relPath := from.Path()
+
+			// Check patterns
+			if in.matchesPatterns(relPath) {
+				select {
+				case in.filesChan <- fileEvent{path: path, isDeleted: true}:
+				case <-in.closeCh:
+					return errors.New("shutdown signaled")
+				}
+			}
+			continue
+		}
+
+		if hasBeenAddedOrModified {
+			path := filepath.Join(in.tempDir, to.Path())
+			relPath := to.Path()
+
+			// Skip directories if not recursive
+			info, err := os.Stat(path)
+			if err != nil {
+				if !os.IsNotExist(err) {
+					in.log.Errorf("Failed to stat file %s: %v", path, err)
+				}
+				continue
+			}
+
+			if info.IsDir() {
+				if !in.cfg.recursive {
+					continue
+				}
+			}
+
+			// Check patterns
+			if in.matchesPatterns(relPath) {
+				select {
+				case in.filesChan <- fileEvent{path: path, isDeleted: false}:
+				case <-in.closeCh:
+					return errors.New("shutdown signaled")
+				}
+			}
+		}
+	}
+
+	// Update the last commit hash
+	in.lastCommit = newCommit
+	return nil
+}
+
+// matchesPatterns checks if the relative path matches the include/exclude patterns.
+func (in *input) matchesPatterns(relPath string) bool {
+	// Check exclude patterns first
+	for _, pattern := range in.cfg.excludePatterns {
+		if matched, err := doublestar.PathMatch(pattern, relPath); err == nil && matched {
+			return false
+		}
+	}
+
+	// If no include patterns, include all files
+	if len(in.cfg.includePatterns) == 0 {
+		return true
+	}
+
+	// Check include patterns
+	for _, pattern := range in.cfg.includePatterns {
+		if matched, err := doublestar.PathMatch(pattern, relPath); err == nil && matched {
+			return true
+		}
+	}
+	return false
+}
+
+// scanFiles walks through the repository directory, applying include/exclude patterns,
+// and sends matching file paths to the files channel for processing.
+func (in *input) scanFiles() error {
+	scanPath := in.tempDir
+
+	return filepath.WalkDir(scanPath, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Skip directories if not recursive
+		if d.IsDir() {
+			if !in.cfg.recursive && path != scanPath {
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		// Get relative path for pattern matching
+		relPath, err := filepath.Rel(scanPath, path)
+		if err != nil {
+			return err
+		}
+
+		// Check patterns
+		if in.matchesPatterns(relPath) {
+			select {
+			case in.filesChan <- fileEvent{path: path, isDeleted: false}:
+			case <-in.closeCh:
+				return errors.New("shutdown signaled")
+			}
+		}
+
+		return nil
+	})
+}
+
+// detectMimeType determines the MIME type of a file by examining its contents or by looking
+// at the file name's extension.
+func (in *input) detectMimeType(filePath string) (string, bool) {
+	// Read the first 512 bytes of the file for MIME detection
+	f, err := os.Open(filePath)
+	if err != nil {
+		in.log.Debugf("failed to open file %q for MIME detection: %v. Using fallback text/plain.", filePath, err)
+		return "unknown/error", false
```

Comment:
Log disagrees. Should we be falling back to `application/octet-stream` and binary?